File: crossing.py
import cv2                 # カメラ画像取得用
import engineio            # リアルタイム通信用
import requests            # HTTP通信用
from ultralytics.models.yolo import YOLO

model = YOLO("yolo11n.pt") # YOLOv11のモデルをロード
cap = cv2.VideoCapture(0)  # カメラデバイスを開く
eio = engineio.Client()    # Engine.IOクライアント生成

@eio.on("connect")
def connect():
    print("connect")
    eio.send("crossing-id test1")  # 踏切IDをサーバに通知

@eio.on("disconnect")
def disconnect():
    print("disconnect")

@eio.on("message")
def message(msg):
    print("message:", msg)
    if msg == "picture request":
        # サーバから撮影リクエストを受けたら画像を撮影
        # cap.read() の戻り値を必ずチェックし、frame が None または空画像 の場合は送信しない
        try:
            # カメラが閉じている場合は再オープンを試みる
            global cap
            if not cap or not cap.isOpened():
                try:
                    cap.release()
                except Exception:
                    pass
                cap = cv2.VideoCapture(0)

            ret, frame = cap.read()
        except Exception as e:
            print("cap.read() エラー:", e)
            return

        # フレームが正しいか厳密にチェック
        if not ret or frame is None:
            print("カメラ読み取り失敗またはフレームが None。画像を送信しません。")
            return
        # フレームが空配列の場合をチェック
        if not hasattr(frame, "size") or frame.size == 0:
            print("フレームが空です。送信を中止します。")
            return
        # 形状チェック（高さ/幅がゼロでないこと）
        try:
            h, w = frame.shape[:2]
            if h == 0 or w == 0:
                print("フレームの幅または高さが0です。送信を中止します。")
                return
        except Exception:
            # shape が取得できない場合も送信中止
            print("フレームの shape を取得できません。送信を中止します。")
            return

        # PNG へエンコードは例外を投げる可能性があるため try/except で保護
        try:
            success, buf = cv2.imencode('.png', frame)
        except cv2.error as e:
            print("cv2.imencode エラー:", e)
            return
        except Exception as e:
            print("画像エンコード時の予期せぬエラー:", e)
            return

        if not success or buf is None:
            print("画像エンコード失敗。送信しません。")
            return

        data = buf.tobytes()

        # 画像データをサーバに送信
        try:
            response = requests.post(
                url="http://localhost:3000/picture",
                data=data,
                params={"crossing-id":"test1"},
                timeout=5
            )
            print("画像データを送信しました. status:", response.status_code)
        except Exception as e:
            print("画像送信エラー:", e)

# サーバに接続
try:
    eio.connect("http://localhost:3000", transports=["polling"])
    eio.wait()  # イベントループ
except Exception as e:
    print("Engine.IO 接続エラー:", e)

File: crossing.rb
require 'faye/websocket'    # WebSocket通信のためのライブラリ
require 'eventmachine'      # 非同期イベント処理ライブラリ
require 'net/http'          # HTTP通信のためのライブラリ
require 'uri'               # URI操作用ライブラリ

EM.run do
  # WebSocketクライアントを作成し、サーバに接続
  ws = Faye::WebSocket::Client.new('ws://localhost:3000/ws')

  # WebSocket接続が確立したときの処理
  ws.on :open do |event|
    ws.send('crossing-id test1')   # crossing-idをサーバに送信して登録
    puts "WebSocket connected"     # 接続完了メッセージ表示
  end

  ws.on :error do  |event|
    p event
    p event.http
    p event.message
    puts "Error: #{event.message}"  # エラーメッセージ表示
  end


  # サーバからメッセージを受信したときの処理
  ws.on :message do |event|
    puts "message: #{event.data}"  # 受信したメッセージを表示
    if event.data == 'picture request'  # サーバから撮影リクエストが来た場合
      buf = File.binread('image.jpg')   # image.jpgファイルをバイナリで読み込む
      uri = URI('http://localhost:3000/picture?crossing-id=test1') # crossing-idをURLに含める
      headers = { 'Content-Type' => 'application/octet-stream' }   # バイナリ送信時はContent-Typeを指定

      Net::HTTP.start(uri.host, uri.port) do |http|

        req = Net::HTTP::Post.new(uri, headers)
        req.body = buf
        res = http.request(req)
        puts res.body
      end

      puts "画像データを送信しました"     # 送信完了メッセージ表示
    end
  end

  # WebSocket接続が切断されたときの処理
  ws.on :close do |event|
    puts "WebSocket closed"         # 切断メッセージ表示
    EM.stop                         # EventMachineのループを停止
  end

end
